lsp-methods.el

```diff
@@ -329,37 +334,69 @@ interface TextDocumentItem {
-  (with-demoted-errors "LSP error: %S"
-    (lsp--send-request (lsp--make-request "shutdown" (make-hash-table)) t)
-    (lsp--send-notification (lsp--make-notification "exit" nil)))
-  (lsp--uninitialize-workspace))
-
 ;; Clean up the entire state of lsp mode when Emacs is killed, to get rid of any
 ;; pending language servers.
 (add-hook 'kill-emacs-hook #'lsp--global-teardown)
 
 (defun lsp--global-teardown ()
   (with-demoted-errors "Error in ‘lsp--global-teardown’: %S"
-    (maphash (lambda (_k value) (lsp--teardown-client value)) lsp--workspaces)))
+    (maphash (lambda (_k value) (lsp--teardown-workspace value)) lsp--workspaces)))
 
-(defun lsp--teardown-client (client)
-  (setq lsp--cur-workspace client)
+(defun lsp--teardown-workspace (workspace)
+  (setq lsp--cur-workspace workspace)
   (lsp--shutdown-cur-workspace))
 
+(defun lsp--shutdown-cur-workspace ()
+  "Shut down the language server process for ‘lsp--cur-workspace’."
+  (with-demoted-errors "LSP error: %S"
+    (lsp--send-request (lsp--make-request "shutdown" (make-hash-table)) t)
+    (lsp--send-notification (lsp--make-notification "exit" nil)))
+  (lsp--uninitialize-workspace))
+
 (defun lsp--uninitialize-workspace ()
   "When a workspace is shut down, by request or from just
 disappearing, unset all the variables related to it."
-  (remhash (lsp--workspace-root lsp--cur-workspace) lsp--workspaces)
-  (let (proc)
+  (let (proc
+        (root (lsp--workspace-root lsp--cur-workspace)))
     (with-current-buffer (current-buffer)
       (setq proc (lsp--workspace-proc lsp--cur-workspace))
       (unless (eq (process-status proc) 'exit)
         (kill-process (lsp--workspace-proc lsp--cur-workspace)))
       (setq lsp--cur-workspace nil)
       (lsp--unset-variables)
-      (kill-local-variable 'lsp--cur-workspace))))
+      (kill-local-variable 'lsp--cur-workspace))
+    (remhash root lsp--workspaces))
+  )
+
+(defun lsp-restart-workspace ()
+  "Shut down and then restart the current workspace.
+This involves uninitializing each of the buffers associated with
+the workspace, closing the process managing communication with
+the client, and then starting up again."
+  (interactive)
+  (when (and (lsp-mode) (buffer-file-name))
+    (let ((old-buffers (lsp--workspace-buffers lsp--cur-workspace))
+          (restart (lsp--client-restart-command (lsp--workspace-client lsp--cur-workspace)))
+          (proc (lsp--workspace-proc lsp--cur-workspace)))
+
+      ;; Shut down the LSP mode for each buffer in the workspace
+      (dolist (buffer1 old-buffers)
+        (with-current-buffer buffer1
+          (lsp--text-document-did-close)
+          (setq lsp--cur-workspace nil)
+          (lsp-mode -1)))
+
+      ; Let the process actually shut down
+      (while (process-live-p proc)
+        (accept-process-output proc))
+
+      ; Re-enable LSP mode for each buffer
+      (dolist (buffer2 old-buffers)
+        (with-current-buffer buffer2
+          (funcall restart)
```

Comment:
Nit: it's more common in Lisp to not insert linebreaks before closing parentheses.